### ISSUE TITLE
Ensure compatibility with Microsoft.EntityFrameworkCore v5+

### ISSUE THREAD

**Issue by AlexBagnolini** [googleapis/dotnet-spanner-entity-framework#83]
## Expected Behavior
Project works as intended.

## Actual Behavior
An exception is thrown during context initialization.

## Steps to Reproduce the Problem

1. Create a new project
2. Add a reference to Google.Cloud.EntityFrameworkCore.Spanner (either nuget or project)
3. Add the Microsoft.EntityFrameworkCore nuget **version 5+** (tested with 5.0.7+)
4. Add any kind of context example
5. Run it

## Exception thrown
```
System.InvalidOperationException
  HResult=0x80131509
  Message=The database provider attempted to register an implementation of the 'IRelationalTypeMappingSource' service. This is not a service defined by Entity Framework and as such must be registered as a provider-specific service using the 'TryAddProviderSpecificServices' method.
  Source=Microsoft.EntityFrameworkCore
  StackTrace:
   at Microsoft.EntityFrameworkCore.Infrastructure.EntityFrameworkServicesBuilder.GetServiceCharacteristics(Type serviceType)
   at Microsoft.EntityFrameworkCore.Infrastructure.EntityFrameworkServicesBuilder.TryAdd(Type serviceType, Type implementationType)
   at Microsoft.EntityFrameworkCore.Infrastructure.EntityFrameworkServicesBuilder.TryAdd[TService,TImplementation]()
   at Google.Cloud.EntityFrameworkCore.Spanner.Extensions.SpannerServiceCollectionExtensions.AddEntityFrameworkSpanner(IServiceCollection serviceCollection) in C:\Source\MySpanneTest\Projects\Google.Cloud.EntityFrameworkCore.Spanner\Extensions\SpannerServiceCollectionExtensions.cs:line 46
   at Google.Cloud.EntityFrameworkCore.Spanner.Infrastructure.Internal.SpannerOptionsExtension.ApplyServices(IServiceCollection services) in C:\Source\MySpanneTest\Projects\Google.Cloud.EntityFrameworkCore.Spanner\Infrastructure\Internal\SpannerOptionsExtension.cs:line 78
   at Microsoft.EntityFrameworkCore.Internal.ServiceProviderCache.ApplyServices(IDbContextOptions options, ServiceCollection services)
   at Microsoft.EntityFrameworkCore.Internal.ServiceProviderCache.<>c__DisplayClass4_0.<GetOrAdd>g__BuildServiceProvider|3()
   at Microsoft.EntityFrameworkCore.Internal.ServiceProviderCache.<>c__DisplayClass4_0.<GetOrAdd>b__2(Int64 k)
   at System.Collections.Concurrent.ConcurrentDictionary`2.GetOrAdd(TKey key, Func`2 valueFactory)
   at Microsoft.EntityFrameworkCore.Internal.ServiceProviderCache.GetOrAdd(IDbContextOptions options, Boolean providerRequired)
   at Microsoft.EntityFrameworkCore.DbContext..ctor(DbContextOptions options)
   at MySpanneTest.Cloud.Authentication.API.Models.AuthenticationContext..ctor(DbContextOptions`1 options) in C:\Source\MySpanneTest\Projects\MySpanneTest.Cloud.Authentication\Models\AuthenticationContext.cs:line 12
   at System.RuntimeMethodHandle.InvokeMethod(Object target, Object[] arguments, Signature sig, Boolean constructor, Boolean wrapExceptions)
   at System.Reflection.RuntimeConstructorInfo.Invoke(BindingFlags invokeAttr, Binder binder, Object[] parameters, CultureInfo culture)
   at Microsoft.Extensions.DependencyInjection.ServiceLookup.CallSiteRuntimeResolver.VisitConstructor(ConstructorCallSite constructorCallSite, RuntimeResolverContext context)
   at Microsoft.Extensions.DependencyInjection.ServiceLookup.CallSiteVisitor`2.VisitCallSiteMain(ServiceCallSite callSite, TArgument argument)
   at Microsoft.Extensions.DependencyInjection.ServiceLookup.CallSiteRuntimeResolver.VisitCache(ServiceCallSite callSite, RuntimeResolverContext context, ServiceProviderEngineScope serviceProviderEngine, RuntimeResolverLock lockType)
   at Microsoft.Extensions.DependencyInjection.ServiceLookup.CallSiteRuntimeResolver.VisitScopeCache(ServiceCallSite singletonCallSite, RuntimeResolverContext context)
   at Microsoft.Extensions.DependencyInjection.ServiceLookup.CallSiteVisitor`2.VisitCallSite(ServiceCallSite callSite, TArgument argument)
   at Microsoft.Extensions.DependencyInjection.ServiceLookup.CallSiteRuntimeResolver.Resolve(ServiceCallSite callSite, ServiceProviderEngineScope scope)
   at Microsoft.Extensions.DependencyInjection.ServiceLookup.DynamicServiceProviderEngine.<>c__DisplayClass1_0.<RealizeService>b__0(ServiceProviderEngineScope scope)
   at Microsoft.Extensions.DependencyInjection.ServiceLookup.ServiceProviderEngine.GetService(Type serviceType, ServiceProviderEngineScope serviceProviderEngineScope)
   at Microsoft.Extensions.DependencyInjection.ServiceLookup.ServiceProviderEngineScope.GetService(Type serviceType)
   at Microsoft.Extensions.DependencyInjection.ActivatorUtilities.GetService(IServiceProvider sp, Type type, Type requiredBy, Boolean isDefaultParameterRequired)
   at Microsoft.AspNetCore.Mvc.Controllers.ControllerActivatorProvider.<>c__DisplayClass4_0.<CreateActivator>b__0(ControllerContext controllerContext)
   at Microsoft.AspNetCore.Mvc.Controllers.ControllerFactoryProvider.<>c__DisplayClass5_0.<CreateControllerFactory>g__CreateController|0(ControllerContext controllerContext)
   at Microsoft.AspNetCore.Mvc.Infrastructure.ControllerActionInvoker.Next(State& next, Scope& scope, Object& state, Boolean& isCompleted)
   at Microsoft.AspNetCore.Mvc.Infrastructure.ControllerActionInvoker.InvokeInnerFilterAsync()

```

**Comment by laljikanjareeya** [comment]
@AlexBagnolini Thanks for the report.
Currently, It only supports `Microsoft.EntityFrameworkCore` version `3.1`.

**Comment by skuruppu** [comment]
@AlexBagnolini is this a major blocker for you or is it possible for you to try to use this implementation with 3.1? We would like to understand the requirements and then consider if we should support v5+.

**Comment by AlexBagnolini** [comment]
We successfully reverted to 3.1, but I guess going forward it will become less and less feasible to force a lower version compatibility.

The problem seems to lie in the way it adds provider services:

https://github.com/cloudspannerecosystem/dotnet-spanner-entity-framework/blob/122db7444af48ce6580d3b8ec235c7b0a220c349/Google.Cloud.EntityFrameworkCore.Spanner/Extensions/SpannerServiceCollectionExtensions.cs#L46

Wouldn't replacing `TryAdd `with `TryAddProviderSpecificServices `be enough for it to work?

**Comment by skuruppu** [comment]
Thanks a lot @AlexBagnolini for the context.

@laljikanjareeya is replacing `TryAdd` the only change that's required or is there more to it?

**Comment by laljikanjareeya** [comment]
@AlexBagnolini `EntityFrameworkServicesBuilder.TryAddProviderSpecificServices` method is used for provider-specific services, not for the Entity Framework Services.
Please refer the original [Documentation](https://docs.microsoft.com/en-us/dotnet/api/microsoft.entityframeworkcore.infrastructure.entityframeworkservicesbuilder.tryaddproviderspecificservices?view=efcore-5.0)

e.g. `ISpannerRelationalConnection` is spanner specific and it's been added by `TryAddProviderSpecificServices`
https://github.com/cloudspannerecosystem/dotnet-spanner-entity-framework/blob/c520460874ac78edd68d3c071cb74e80575ac9ae/Google.Cloud.EntityFrameworkCore.Spanner/Extensions/SpannerServiceCollectionExtensions.cs#L67-L68

I am getting different error while performing above mentioned steps and that is because of Microsoft.EntityFrameworkCore 5.0.0 changes.

```
System.TypeInitializationException: The type initializer for 'Google.Cloud.EntityFrameworkCore.Spanner.Storage.Internal.SpannerTypeMappingSource' threw an exception.
 ---> System.MissingMethodException: Method not found: 'Void Microsoft.EntityFrameworkCore.Storage.RelationalTypeMapping..ctor(System.String, System.Type, System.Nullable`1<System.Data.DbType>, Boolean, System.Nullable`1<Int32>)'.
   at Google.Cloud.EntityFrameworkCore.Spanner.Storage.Internal.SpannerTimestampTypeMapping..ctor()
   at Google.Cloud.EntityFrameworkCore.Spanner.Storage.Internal.SpannerTypeMappingSource..cctor() in F:\Google\dotnet\dotnet-spanner-entity-framework\Google.Cloud.EntityFrameworkCore.Spanner\Storage\Internal\SpannerTypeMappingSource.cs:line 37
   --- End of inner exception stack trace ---
   at Google.Cloud.EntityFrameworkCore.Spanner.Storage.Internal.SpannerTypeMappingSource..ctor(TypeMappingSourceDependencies dependencies, RelationalTypeMappingSourceDependencies relationalDependencies) in F:\Google\dotnet\dotnet-spanner-entity-framework\Google.Cloud.EntityFrameworkCore.Spanner\Storage\Internal\SpannerTypeMappingSource.cs:line 164
   at System.RuntimeMethodHandle.InvokeMethod(Object target, Object[] arguments, Signature sig, Boolean constructor, Boolean wrapExceptions)
   at System.Reflection.RuntimeConstructorInfo.Invoke(BindingFlags invokeAttr, Binder binder, Object[] parameters, CultureInfo culture)
   at Microsoft.Extensions.DependencyInjection.ServiceLookup.CallSiteRuntimeResolver.VisitConstructor(ConstructorCallSite constructorCallSite, RuntimeResolverContext context)
   at Microsoft.Extensions.DependencyInjection.ServiceLookup.CallSiteVisitor`2.VisitCallSiteMain(ServiceCallSite callSite, TArgument argument)
   at Microsoft.Extensions.DependencyInjection.ServiceLookup.CallSiteRuntimeResolver.VisitCache(ServiceCallSite callSite, RuntimeResolverContext context, ServiceProviderEngineScope serviceProviderEngine, RuntimeResolverLock lockType)
   at Microsoft.Extensions.DependencyInjection.ServiceLookup.CallSiteRuntimeResolver.VisitRootCache(ServiceCallSite singletonCallSite, RuntimeResolverContext context)
   at Microsoft.Extensions.DependencyInjection.ServiceLookup.CallSiteVisitor`2.VisitCallSite(ServiceCallSite callSite, TArgument argument)
   at Microsoft.Extensions.DependencyInjection.ServiceLookup.CallSiteRuntimeResolver.Resolve(ServiceCallSite callSite, ServiceProviderEngineScope scope)
   at Microsoft.Extensions.DependencyInjection.ServiceLookup.DynamicServiceProviderEngine.<>c__DisplayClass1_0.<RealizeService>b__0(ServiceProviderEngineScope scope)
   at Microsoft.Extensions.DependencyInjection.ServiceLookup.ServiceProviderEngine.GetService(Type serviceType, ServiceProviderEngineScope serviceProviderEngineScope)
   at Microsoft.Extensions.DependencyInjection.ServiceLookup.ServiceProviderEngineScope.GetService(Type serviceType)
   at Microsoft.Extensions.DependencyInjection.ServiceProviderServiceExtensions.GetService[T](IServiceProvider provider)
   at Microsoft.EntityFrameworkCore.Infrastructure.EntityFrameworkRelationalServicesBuilder.<>c.<TryAddCoreServices>b__3_0(IServiceProvider p)
   at Microsoft.Extensions.DependencyInjection.ServiceLookup.CallSiteRuntimeResolver.VisitFactory(FactoryCallSite factoryCallSite, RuntimeResolverContext context)
   at Microsoft.Extensions.DependencyInjection.ServiceLookup.CallSiteVisitor`2.VisitCallSiteMain(ServiceCallSite callSite, TArgument argument)
   at Microsoft.Extensions.DependencyInjection.ServiceLookup.CallSiteRuntimeResolver.VisitCache(ServiceCallSite callSite, RuntimeResolverContext context, ServiceProviderEngineScope serviceProviderEngine, RuntimeResolverLock lockType)
   at Microsoft.Extensions.DependencyInjection.ServiceLookup.CallSiteRuntimeResolver.VisitRootCache(ServiceCallSite singletonCallSite, RuntimeResolverContext context)
   at Microsoft.Extensions.DependencyInjection.ServiceLookup.CallSiteVisitor`2.VisitCallSite(ServiceCallSite callSite, TArgument argument)
   at Microsoft.Extensions.DependencyInjection.ServiceLookup.CallSiteRuntimeResolver.VisitConstructor(ConstructorCallSite constructorCallSite, RuntimeResolverContext context)
   at Microsoft.Extensions.DependencyInjection.ServiceLookup.CallSiteVisitor`2.VisitCallSiteMain(ServiceCallSite callSite, TArgument argument)
   at Microsoft.Extensions.DependencyInjection.ServiceLookup.CallSiteRuntimeResolver.VisitCache(ServiceCallSite callSite, RuntimeResolverContext context, ServiceProviderEngineScope serviceProviderEngine, RuntimeResolverLock lockType)
   at Microsoft.Extensions.DependencyInjection.ServiceLookup.CallSiteRuntimeResolver.VisitScopeCache(ServiceCallSite singletonCallSite, RuntimeResolverContext context)
   at Microsoft.Extensions.DependencyInjection.ServiceLookup.CallSiteVisitor`2.VisitCallSite(ServiceCallSite callSite, TArgument argument)
   at Microsoft.Extensions.DependencyInjection.ServiceLookup.CallSiteRuntimeResolver.VisitConstructor(ConstructorCallSite constructorCallSite, RuntimeResolverContext context)
   at Microsoft.Extensions.DependencyInjection.ServiceLookup.CallSiteVisitor`2.VisitCallSiteMain(ServiceCallSite callSite, TArgument argument)
   at Microsoft.Extensions.DependencyInjection.ServiceLookup.CallSiteRuntimeResolver.VisitCache(ServiceCallSite callSite, RuntimeResolverContext context, ServiceProviderEngineScope serviceProviderEngine, RuntimeResolverLock lockType)
   at Microsoft.Extensions.DependencyInjection.ServiceLookup.CallSiteRuntimeResolver.VisitScopeCache(ServiceCallSite singletonCallSite, RuntimeResolverContext context)
   at Microsoft.Extensions.DependencyInjection.ServiceLookup.CallSiteVisitor`2.VisitCallSite(ServiceCallSite callSite, TArgument argument)
   at Microsoft.Extensions.DependencyInjection.ServiceLookup.CallSiteRuntimeResolver.VisitConstructor(ConstructorCallSite constructorCallSite, RuntimeResolverContext context)
   at Microsoft.Extensions.DependencyInjection.ServiceLookup.CallSiteVisitor`2.VisitCallSiteMain(ServiceCallSite callSite, TArgument argument)
   at Microsoft.Extensions.DependencyInjection.ServiceLookup.CallSiteRuntimeResolver.VisitCache(ServiceCallSite callSite, RuntimeResolverContext context, ServiceProviderEngineScope serviceProviderEngine, RuntimeResolverLock lockType)
   at Microsoft.Extensions.DependencyInjection.ServiceLookup.CallSiteRuntimeResolver.VisitScopeCache(ServiceCallSite singletonCallSite, RuntimeResolverContext context)
   at Microsoft.Extensions.DependencyInjection.ServiceLookup.CallSiteVisitor`2.VisitCallSite(ServiceCallSite callSite, TArgument argument)
   at Microsoft.Extensions.DependencyInjection.ServiceLookup.CallSiteRuntimeResolver.VisitConstructor(ConstructorCallSite constructorCallSite, RuntimeResolverContext context)
   at Microsoft.Extensions.DependencyInjection.ServiceLookup.CallSiteVisitor`2.VisitCallSiteMain(ServiceCallSite callSite, TArgument argument)
   at Microsoft.Extensions.DependencyInjection.ServiceLookup.CallSiteRuntimeResolver.VisitCache(ServiceCallSite callSite, RuntimeResolverContext context, ServiceProviderEngineScope serviceProviderEngine, RuntimeResolverLock lockType)
   at Microsoft.Extensions.DependencyInjection.ServiceLookup.CallSiteRuntimeResolver.VisitScopeCache(ServiceCallSite singletonCallSite, RuntimeResolverContext context)
   at Microsoft.Extensions.DependencyInjection.ServiceLookup.CallSiteVisitor`2.VisitCallSite(ServiceCallSite callSite, TArgument argument)
   at Microsoft.Extensions.DependencyInjection.ServiceLookup.CallSiteRuntimeResolver.Resolve(ServiceCallSite callSite, ServiceProviderEngineScope scope)
   at Microsoft.Extensions.DependencyInjection.ServiceLookup.DynamicServiceProviderEngine.<>c__DisplayClass1_0.<RealizeService>b__0(ServiceProviderEngineScope scope)
   at Microsoft.Extensions.DependencyInjection.ServiceLookup.ServiceProviderEngine.GetService(Type serviceType, ServiceProviderEngineScope serviceProviderEngineScope)
   at Microsoft.Extensions.DependencyInjection.ServiceLookup.ServiceProviderEngineScope.GetService(Type serviceType)
   at Microsoft.Extensions.DependencyInjection.ServiceProviderServiceExtensions.GetService[T](IServiceProvider provider)
   at Microsoft.EntityFrameworkCore.Internal.DbContextServices.CreateModel()
   at Microsoft.EntityFrameworkCore.Internal.DbContextServices.get_Model()
   at Microsoft.EntityFrameworkCore.Infrastructure.EntityFrameworkServicesBuilder.<>c.<TryAddCoreServices>b__7_3(IServiceProvider p)
   at Microsoft.Extensions.DependencyInjection.ServiceLookup.CallSiteRuntimeResolver.VisitFactory(FactoryCallSite factoryCallSite, RuntimeResolverContext context)
   at Microsoft.Extensions.DependencyInjection.ServiceLookup.CallSiteVisitor`2.VisitCallSiteMain(ServiceCallSite callSite, TArgument argument)
   at Microsoft.Extensions.DependencyInjection.ServiceLookup.CallSiteRuntimeResolver.VisitCache(ServiceCallSite callSite, RuntimeResolverContext context, ServiceProviderEngineScope serviceProviderEngine, RuntimeResolverLock lockType)
   at Microsoft.Extensions.DependencyInjection.ServiceLookup.CallSiteRuntimeResolver.VisitScopeCache(ServiceCallSite singletonCallSite, RuntimeResolverContext context)
   at Microsoft.Extensions.DependencyInjection.ServiceLookup.CallSiteVisitor`2.VisitCallSite(ServiceCallSite callSite, TArgument argument)
   at Microsoft.Extensions.DependencyInjection.ServiceLookup.CallSiteRuntimeResolver.VisitConstructor(ConstructorCallSite constructorCallSite, RuntimeResolverContext context)
   at Microsoft.Extensions.DependencyInjection.ServiceLookup.CallSiteVisitor`2.VisitCallSiteMain(ServiceCallSite callSite, TArgument argument)
   at Microsoft.Extensions.DependencyInjection.ServiceLookup.CallSiteRuntimeResolver.VisitCache(ServiceCallSite callSite, RuntimeResolverContext context, ServiceProviderEngineScope serviceProviderEngine, RuntimeResolverLock lockType)
   at Microsoft.Extensions.DependencyInjection.ServiceLookup.CallSiteRuntimeResolver.VisitScopeCache(ServiceCallSite singletonCallSite, RuntimeResolverContext context)
   at Microsoft.Extensions.DependencyInjection.ServiceLookup.CallSiteVisitor`2.VisitCallSite(ServiceCallSite callSite, TArgument argument)
   at Microsoft.Extensions.DependencyInjection.ServiceLookup.CallSiteRuntimeResolver.Resolve(ServiceCallSite callSite, ServiceProviderEngineScope scope)
   at Microsoft.Extensions.DependencyInjection.ServiceLookup.DynamicServiceProviderEngine.<>c__DisplayClass1_0.<RealizeService>b__0(ServiceProviderEngineScope scope)
   at Microsoft.Extensions.DependencyInjection.ServiceLookup.ServiceProviderEngine.GetService(Type serviceType, ServiceProviderEngineScope serviceProviderEngineScope)
   at Microsoft.Extensions.DependencyInjection.ServiceLookup.ServiceProviderEngineScope.GetService(Type serviceType)
   at Microsoft.Extensions.DependencyInjection.ServiceProviderServiceExtensions.GetRequiredService(IServiceProvider provider, Type serviceType)
   at Microsoft.Extensions.DependencyInjection.ServiceProviderServiceExtensions.GetRequiredService[T](IServiceProvider provider)
   at Microsoft.EntityFrameworkCore.DbContext.get_DbContextDependencies()
   at Microsoft.EntityFrameworkCore.DbContext.get_InternalServiceProvider()
   at Microsoft.EntityFrameworkCore.DbContext.Microsoft.EntityFrameworkCore.Infrastructure.IInfrastructure<System.IServiceProvider>.get_Instance()
   at Microsoft.EntityFrameworkCore.Infrastructure.Internal.InfrastructureExtensions.GetService[TService](IInfrastructure`1 accessor)
   at Microsoft.EntityFrameworkCore.Infrastructure.AccessorExtensions.GetService[TService](IInfrastructure`1 accessor)
   at Microsoft.EntityFrameworkCore.Design.Internal.DbContextOperations.CreateContext(Func`1 factory)
   at Microsoft.EntityFrameworkCore.Design.Internal.DbContextOperations.CreateContext(String contextType)
   at Microsoft.EntityFrameworkCore.Design.Internal.MigrationsOperations.AddMigration(String name, String outputDir, String contextType, String namespace)
   at Microsoft.EntityFrameworkCore.Design.OperationExecutor.AddMigrationImpl(String name, String outputDir, String contextType, String namespace)
   at Microsoft.EntityFrameworkCore.Design.OperationExecutor.AddMigration.<>c__DisplayClass0_0.<.ctor>b__0()
   at Microsoft.EntityFrameworkCore.Design.OperationExecutor.OperationBase.<>c__DisplayClass3_0`1.<Execute>b__0()
   at Microsoft.EntityFrameworkCore.Design.OperationExecutor.OperationBase.Execute(Action action)
The type initializer for 'Google.Cloud.EntityFrameworkCore.Spanner.Storage.Internal.SpannerTypeMappingSource' threw an exception.
```
Could you explain on which step you are getting above exception as I have tried with both approaches i.e. [Database Migrations](https://github.com/cloudspannerecosystem/dotnet-spanner-entity-framework#database-migrations) and [Database-First](https://github.com/cloudspannerecosystem/dotnet-spanner-entity-framework#create-model-for-an-existing-database-database-first--scaffolding) but both throws above exception?

**Comment by AlexBagnolini** [comment]
Hi, I just tried again updating the EF Core package to 5.0.7 and I'm getting the previous exception I reported:

`System.InvalidOperationException: 'The database provider attempted to register an implementation of the 'IRelationalTypeMappingSource' service. This is not a service defined by Entity Framework and as such must be registered as a provider-specific service using the 'TryAddProviderSpecificServices' method.'`

Here is the package list I'm using in my project:
```
  <ItemGroup>
    <PackageReference Include="Grpc.Core" Version="2.38.1" />
    <PackageReference Include="Google.Cloud.EntityFrameworkCore.Spanner" Version="1.0.0-beta" />
    <PackageReference Include="Microsoft.AspNetCore.Authentication.JwtBearer" Version="5.0.7" />
    <PackageReference Include="Microsoft.AspNetCore.Mvc.Versioning" Version="5.0.0" />
    <PackageReference Include="Microsoft.EntityFrameworkCore" Version="5.0.7" />
    <PackageReference Include="Microsoft.VisualStudio.Azure.Containers.Tools.Targets" Version="1.10.14" />
    <PackageReference Include="Microsoft.VisualStudio.Azure.Kubernetes.Tools.Targets" Version="1.1.0" />
    <PackageReference Include="System.IdentityModel.Tokens.Jwt" Version="6.11.1" />
  </ItemGroup>
```
Please note that the `Google.Cloud.EntityFrameworkCore.Spanner` package is just a nuget compiled version of the current master code.

I'm using a database first approach, no migration used.

Let me know if you need anything else!

**Comment by laljikanjareeya** [comment]
@AlexBagnolini you have not scaffolded your model using [scaffolding](https://github.com/cloudspannerecosystem/dotnet-spanner-entity-framework#create-model-for-an-existing-database-database-first--scaffolding) instead you have written it manually. if you do database first scaffolding it will throw the above mention error.

Currently `Google.Cloud.EntityFrameworkCore.Spanner` is implementing the `Microsoft.EntityFrameworkCore.Relational` version `3.1.0` while `Microsoft.EntityFrameworkCore`  is using higher version.

[Here](https://stackoverflow.com/questions/65422945/cannot-register-irelationaltypemappingsource) is the same type of issue reported on stackoverflow.

If you try to update `Microsoft.EntityFrameworkCore.Relational` into your `Google.Cloud.EntityFrameworkCore.Spanner` project it will give you lots of errors.

@skuruppu it's required to update the `Microsoft.EntityFrameworkCore.Relational` package to support the `Microsoft.EntityFrameworkCore` version `5+`.

**Comment by skuruppu** [comment]
@laljikanjareeya do you know what the scope is for updating the `Microsoft.EntityFrameworkCore.Relational` package? You can reply to me offline if that makes more sense.

**Comment by AlexBagnolini** [comment]
@laljikanjareeya You were right, changing the Relational and Design versions to 5.0.7 made the package compatible with EFCore 5.

Doing so required some changes though, the main one being having to upgrade to netstandard 2.1 (up from 2.0).
A lot of Relational classes changed too, and required changes to many classes to make it compatible.

If you find it of any usefulness, I made a commit in my fork here: https://github.com/AlexBagnolini/dotnet-spanner-entity-framework/commit/b9104dfda30c0a9c54ac0fa5f5f20a00af33f152

I didn't want to venture too much into changing SpannerMigrationsAnnotationProvider so I just commented it out as I don't use Migrations.

Thanks for all your support.

**Comment by chrisbbe** [comment]
EF Core 6 is the latest LTS, any chance you will upgrade EF Core to version 6 shortly? We are currently doing a PoC in an already existing application using EF Core 6.

**Comment by olavloite** [comment]
@chrisbbe I had a quick look today to see what would be required to support EF Core 6, but it is not trivial. This is partly caused by the fact that EF Core 6 [requires .NET 6](https://docs.microsoft.com/en-us/ef/core/miscellaneous/platforms), which is not supported by `Grpc.Core` on Mac OS. So we would need to add some way to use the [workaround specified here](https://github.com/googleapis/google-cloud-dotnet/issues/7560#issuecomment-975414370) for EF Core for those using Mac OS.
I'll try to look into it a little bit more in detail tomorrow to see if I at least can get a test setup up and running, and get back to you.

**Comment by olavloite** [comment]
@chrisbbe #152 Now contains a version that compiles and successfully runs all tests using EF Core 6.0. You could have a look and see if that is enough for you to at least do a small PoC. As mentioned above, one limitation is that the gRPC libraries that are currently used by the Spanner client library are not compatible with Mac M1 machines, so if any of your developers use that, you would need to apply some additional workarounds to get it up and running. I'm happy to supply an example for that as well if that is necessary.

**Comment by ansh0l** [comment]
Hi @shivgautam : It looks like EF core 3.1 has LTS support till Dec' 2022. (https://docs.microsoft.com/en-us/ef/core/what-is-new/). Please take a look at this issue and associated PR from @olavloite .

**Comment by olavloite** [comment]
We have released version 2.0.0 to NuGet which supports EF 6.0 and .NET 6.0: https://www.nuget.org/packages/Google.Cloud.EntityFrameworkCore.Spanner/2.0.0
